frame/system/src/lib.rs

```diff
@@ -923,6 +928,44 @@ impl<T: Trait> Module<T> {
 	}
 }
 
+impl<T: Trait> Module<T> {
+	pub fn spent_weight() -> Option<Weight> {
+		CurrentDispatchableWeight::get()
+	}
+
+	/// Decrease the amount of weight consumed by the current dispatchable.
+	///
+	/// This function can only meaningfully called within `Self::dispatch`. Otherwise,
+	/// it does nothing and returns `Err`.
+	///
+	/// The total weight returned from the dispatchable cannot exceed its total requested weight.
+	pub fn return_unspent_weight(unspent: Weight) -> Result<(), ()> {
+		let weight = match CurrentDispatchableWeight::take() {
+			None => return Err(()),
+			Some(weight) => weight,
+		};
+		let next_weight = weight.checked_sub(unspent).ok_or(())?;
+		CurrentDispatchableWeight::put(next_weight);
+		Ok(())
+	}
+
+	/// Dispatch a given `T::Call` with a given origin.
+	///
+	/// Returns how much weight was unspent.
+	pub fn dispatch_call(
+		origin: <T::Call as Dispatchable>::Origin,
+		call: T::Call,
+	) -> (Weight, sp_runtime::DispatchResult) {
+		let original_weight = call.get_dispatch_info().weight;
+		CurrentDispatchableWeight::put(original_weight);
+		let result = call.dispatch(origin);
+		// TODO: This won't work with the nested calls.
```

Comment:
didn't we come up with a way in out call that it did (I have to watch it again though.. damn memory)

Reply:
Actually we came up! I assumed that this should be trivial to solve: we preserve the value of `CurrentDispatchableWeight` before and restore it after the dispatch.